handler-ssl-ocsp/src/main/java/io/netty/handler/ssl/ocsp/OcspClient.java

```diff
@@ -0,0 +1,311 @@
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateHolder;
+import org.bouncycastle.cert.ocsp.BasicOCSPResp;
+import org.bouncycastle.cert.ocsp.CertificateID;
+import org.bouncycastle.cert.ocsp.OCSPException;
+import org.bouncycastle.cert.ocsp.OCSPReqBuilder;
+import org.bouncycastle.cert.ocsp.OCSPResp;
+import org.bouncycastle.operator.ContentVerifierProvider;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaContentVerifierProviderBuilder;
+import org.bouncycastle.operator.jcajce.JcaDigestCalculatorProviderBuilder;
+
+import java.net.InetSocketAddress;
+import java.net.URL;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.X509Certificate;
+
+import static io.netty.handler.codec.http.HttpMethod.POST;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+import static io.netty.handler.ssl.ocsp.OcspHttpHandler.OCSP_REQUEST_TYPE;
+import static io.netty.handler.ssl.ocsp.OcspHttpHandler.OCSP_RESPONSE_TYPE;
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static org.bouncycastle.asn1.ocsp.OCSPObjectIdentifiers.id_pkix_ocsp_nonce;
+import static org.bouncycastle.asn1.x509.X509ObjectIdentifiers.id_ad_ocsp;
+import static org.bouncycastle.cert.ocsp.CertificateID.HASH_SHA1;
+
+final class OcspClient {
+
+    /**
+     * Query the certificate status using OCSP
+     *
+     * @param eventLoop             {@link EventLoop} for HTTP communication
+     * @param x509Certificate       Client {@link X509Certificate} to validate
+     * @param issuer                {@link X509Certificate} issuer of client certificate
+     * @param validateResponseNonce Set to {@code true} to enable OCSP response validation
+     * @return {@link Promise} of {@link BasicOCSPResp}
+     */
+    static Promise<BasicOCSPResp> query(final EventLoop eventLoop, final X509Certificate x509Certificate,
+                                 final X509Certificate issuer, final boolean validateResponseNonce) {
+        final Promise<BasicOCSPResp> responsePromise = eventLoop.newPromise();
+        eventLoop.execute(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    CertificateID certificateID = new CertificateID(new JcaDigestCalculatorProviderBuilder()
+                            .build().get(HASH_SHA1), new JcaX509CertificateHolder(issuer),
+                            x509Certificate.getSerialNumber());
+
+                    // Initialize OCSP Request Builder and add CertificateID into it.
+                    OCSPReqBuilder builder = new OCSPReqBuilder();
+                    builder.addRequest(certificateID);
+
+                    // Generate 6-bytes of nonce and add it into OCSP Request builder.
+                    byte[] nonce = new byte[6];
+                    ThreadLocalRandom.current().nextBytes(nonce);
```

Comment:
This needs to use `SecureRandom` (cache a static instance because seeding those can be expensive).

See https://www.rfc-editor.org/rfc/rfc8954.html#name-nonce-collision